File: CodingInterviewsPackSolutionsCPP/Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int>charToIndex;
        int sz = s.size();
        int result = 0;
        for(int start = 0, end = 0; end < sz; end++){
            if(charToIndex.count(s[end]))
                start = max(start, charToIndex[s[end]]+1);
            charToIndex[s[end]] = end;
            result = max(result,end-start+1);
        }
        return result;
    }
};

File: CodingInterviewsJavaSolutions/Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> charToIndex = new HashMap<>();
        int sz = s.length();
        int result = 0;
        int start = 0;
        for (int end = 0; end < sz; end++) {
            if (charToIndex.containsKey(s.charAt(end))) {
                start = Math.max(start, charToIndex.get(s.charAt(end)) + 1);
            }
            charToIndex.put(s.charAt(end), end);
            result = Math.max(result, end - start + 1);
        }
        return result;
    }
}
